File: iosApp/iosApp/Pages/More/MoreItem.swift
//
//  MoreItem.swift
//  iosApp
//
//  Created by esimon on 10/28/24.
//  Copyright © 2024 MBTA. All rights reserved.
//

import Foundation
import Shared

enum MoreItem: Identifiable, Equatable {
    case toggle(label: String, setting: Settings)
    case link(label: String, url: String, note: String? = nil)
    case phone(label: String, phoneNumber: String)
    case navLink(label: String, destination: MoreNavTarget)

    var id: String {
        switch self {
        case let .toggle(label: _, setting: setting): setting.name
        case let .link(label: _, url: url, note: _): url
        case let .phone(label: _, phoneNumber: number): number
        case let .navLink(label: _, destination: destination): destination.id
        }
    }
}

File: shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/morePage/MoreItem.kt
package com.mbta.tid.mbta_app.model.morePage

import com.mbta.tid.mbta_app.repositories.Settings
import com.mbta.tid.mbta_app.utils.SharedString

public sealed class MoreItem {

    public data class Action internal constructor(val label: SharedString, val action: () -> Unit) :
        MoreItem()

    public data class Link(
        val label: SharedString,
        val url: String,
        val note: SharedString? = null,
    ) : MoreItem()

    public data class NavLink(
        val label: SharedString,
        val callback: () -> Unit,
        val note: String? = null,
    ) : MoreItem()

    public data class Phone(val label: String, val phoneNumber: String) : MoreItem()

    public data class Toggle(val label: SharedString, val settings: Settings) : MoreItem()

    internal fun id() {
        when (this) {
            is Action -> label
            is Link -> url
            is NavLink -> label
            is Phone -> phoneNumber
            is Toggle -> settings.name
        }
    }
}
